aptos-move/move-examples/post_mint_reveal_nft/sources/whitelist.move

```diff
@@ -0,0 +1,128 @@
+            if (whitelist_stage.whitelist_minting_start_time <= now && now < whitelist_stage.whitelist_minting_end_time) {
+                return (true, i)
+            };
+            i = i + 1;
+        };
+        return (false, 0)
+    }
+
+    #[view]
+    /// Returns the number of total stages available.
+    public fun get_num_of_stages(module_address: address): u64 acquires WhitelistMintConfig {
+        vector::length(&borrow_global<WhitelistMintConfig>(module_address).whitelist_configs)
+    }
+
+    #[view]
+    /// Checks if the given address is whitelisted in the specified whitelist stage.
+    /// Return value u64 is the minting price of the current stage.
+    /// Return value bool is true if the given address is whitelisted.
+    public fun is_user_eligible_for_minting(module_address: address, minter_address: address, stage: u64): (u64, bool) acquires WhitelistMintConfig {
+        let whitelist_mint_config = borrow_global_mut<WhitelistMintConfig>(module_address);
+        assert!(stage < vector::length(&whitelist_mint_config.whitelist_configs), error::invalid_argument(EINVALID_STAGE));
+        let whitelist_stage = vector::borrow_mut(&mut whitelist_mint_config.whitelist_configs, stage);
+        let user_address_on_whitelist = bucket_table::contains(&whitelist_stage.whitelisted_address, &minter_address);
+        (whitelist_stage.whitelist_mint_price, user_address_on_whitelist)
+    }
+
+    /// Initializes the WhitelistMintConfig resource.
+    public fun init_whitelist_config(admin: &signer) {
+        let config = WhitelistMintConfig {
+            whitelist_configs: vector::empty<WhitelistStage>(),
+        };
+        move_to(admin, config);
+    }
+
+    /// Deducts the number of source certificate that the user wants to mint from the user's minting limit.
+    public(friend) fun deduct_user_minting_amount(module_address: address, minter_address: address, stage: u64, user_minting_amount: u64) acquires WhitelistMintConfig {
+        let whitelist_mint_config = borrow_global_mut<WhitelistMintConfig>(module_address);
+        assert!(stage < vector::length(&whitelist_mint_config.whitelist_configs), error::invalid_argument(EINVALID_STAGE));
+        let whitelist_stage = vector::borrow_mut(&mut whitelist_mint_config.whitelist_configs, stage);
+        assert!(bucket_table::contains(&whitelist_stage.whitelisted_address, &minter_address), error::permission_denied(EACCOUNT_NOT_WHITELISTED));
+        let remaining_minting_amount = bucket_table::borrow_mut(&mut whitelist_stage.whitelisted_address, minter_address);
+        assert!(*remaining_minting_amount >= user_minting_amount, error::invalid_argument(EEXCEEDS_MINT_LIMIT));
+        *remaining_minting_amount = *remaining_minting_amount - user_minting_amount;
+    }
+
+    /// Adds a new whitelist stage.
+    public fun add_whitelist_stage(admin: &signer, whitelist_start_time: u64, whitelist_end_time: u64, whitelist_price: u64, stage: u64) acquires WhitelistMintConfig {
```

Comment:
Why does user needs to pass stage here if it's always equal to the length of whitelist_configs?

Reply:
ah yeah good point, I was planning to make it possible to overwrite but realized it wasn't possible with the current non-droppable table scheme, but forgot to remove the stage index

Reply:
yeah good point - just updated this to make sure a user can update / overwrite an existing whitelist stage